pkg/controller/controller.go

```diff
@@ -1124,6 +1129,24 @@ func (p *csiProvisioner) getSnapshotSource(ctx context.Context, claim *v1.Persis
 		}
 	}
 
+	if p.preventVolumeModeConversion {
+		if snapContentObj.Spec.SourceVolumeMode != nil && snapContentObj.Spec.SourceVolumeMode != claim.Spec.VolumeMode {
+			// Attempt to modify volume mode during volume creation.
+			// Verify if this volume is allowed to alter its mode.
+			allowVolumeModeChange, ok := snapContentObj.Annotations[annAllowVolumeModeChange]
+			if !ok {
+				return nil, fmt.Errorf("requested volume %s/%s modifies the mode of the source volume but does not have permission to do so.", claim.Namespace, claim.Name)
+			}
+			allowVolumeModeChangeBool, err := strconv.ParseBool(allowVolumeModeChange)
+			if err != nil {
+				return nil, fmt.Errorf("failed to convert %s annotation value to boolean", annAllowVolumeModeChange)
+			}
+			if !allowVolumeModeChangeBool {
+				return nil, fmt.Errorf("requested volume %s/%s modifies the mode of the source volume but does not have permission to do so.", claim.Namespace, claim.Name)
```

Comment:
Also add that this annotation is not set to true.

Reply:
Fixed